### DISCUSSION TITLE
[JS201] 日常生活中的作用域陷阱

### DISCUSSION THREAD

**Discussion by spp72kimo** [Lidemy/forum#271] [其他問題] [answered]
老師、同學好
這堂課老師裡面的範例是
```js
var a = [];
for(var i=0; i<5; i++) {
    a[i] = function() {
        console.log(i);
    }
}
a[0]();
```

老師後來用了 Closure 的方式來處理  var 的問題。
我後來想說能不能直接傳當下的參數到 function 裡面，來解決 var 的問題，但是最後 console.log 卻是 undefined，
我想說 function 的參數會優先 Hoisting，並且若變數名稱相同，會以 function 裡面的 parameter value 為主...
不太知道問甚麼會顯示 undefined ，希望老師或是前輩能幫我解答，感謝！！（程式碼如下）
```js
var a = [];
for(var i=0; i<5; i++) {
    a[i] = function(i) {  // 這邊改成傳當下的 i 進去
        console.log(i);
    }
}
a[0]();  // 不太懂為什麼這邊會顯示 undefined
```

**Comment by cwc329** [comment, marked as answer]
```js
a[i] = function(i) { console.log(i) }
```
a[i] 的賦值是一個 function 沒錯，
但是這樣的宣告方式是這個 function 需要傳一個參數 `i`，
然後 function 才會 log 出這個 `i`。
下面這一段和上面是等價的。
```js
a[i] = function(j) { console.log(j) }
```

**Reply by spp72kimo** [reply]
喔喔！是我沒有傳參數的部分><...
感謝指正^^
